src/responder/pam/pamsrv_gssapi.c

```diff
@@ -0,0 +1,594 @@
+    gss_release_buffer(&minor, &buf);
+
+    if (exported == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Out of memory!\n");
+        return NULL;
+    }
+
+    return exported;
+}
+
+struct gssapi_state {
+    bool established;
+    gss_ctx_id_t ctx;
+};
+
+int gssapi_state_destructor(struct gssapi_state *state)
+{
+    OM_uint32 minor;
+
+    gss_delete_sec_context(&minor, &state->ctx, NULL);
+
+    return 0;
+}
+
+static struct gssapi_state * gssapi_get_state(struct cli_ctx *cli_ctx)
+{
+    struct gssapi_state *state;
+
+    state = talloc_get_type(cli_ctx->state_ctx, struct gssapi_state);
+    if (state == NULL) {
+        state = talloc_zero(cli_ctx, struct gssapi_state);
+        if (state == NULL) {
+            return NULL;
+        }
+
+        state->ctx = GSS_C_NO_CONTEXT;
+        talloc_set_destructor(state, gssapi_state_destructor);
+
+        cli_ctx->state_ctx = state;
+    }
+
+    return state;
+}
+
+static errno_t gssapi_get_creds(const char *keytab,
+                                const char *target,
+                                gss_cred_id_t *_creds)
+{
+    gss_key_value_set_desc cstore = {0, NULL};
+    gss_key_value_element_desc el;
+    gss_buffer_desc name_buf;
+    gss_name_t name = GSS_C_NO_NAME;
+    OM_uint32 major;
+    OM_uint32 minor;
+    errno_t ret;
+
+    if (keytab != NULL) {
+        el.key = "keytab";
+        el.value = keytab;
+        cstore.count = 1;
+        cstore.elements = &el;
+    }
+
+    if (target != NULL) {
+        name_buf.value = discard_const(target);
+        name_buf.length = strlen(target);
+
+        major = gss_import_name(&minor, &name_buf, GSS_C_NT_HOSTBASED_SERVICE,
+                                &name);
+        if (GSS_ERROR(major)) {
+            DEBUG(SSSDBG_OP_FAILURE, "Could not import name [%s] "
+                 "[maj:0x%x, min:0x%x]\n", target, major, minor);
+
+            gssapi_log_status(GSS_C_GSS_CODE, major);
+            gssapi_log_status(GSS_C_MECH_CODE, minor);
+
+            ret = EIO;
+            goto done;
+        }
+    }
+
+    major = gss_acquire_cred_from(&minor, name, GSS_C_INDEFINITE,
+                                  GSS_C_NO_OID_SET, GSS_C_ACCEPT, &cstore,
+                                  _creds, NULL, NULL);
+    if (GSS_ERROR(major)) {
+        DEBUG(SSSDBG_OP_FAILURE, "Unable to read credentials from [%s] "
+              "[maj:0x%x, min:0x%x]\n", keytab, major, minor);
+
+        gssapi_log_status(GSS_C_GSS_CODE, major);
+        gssapi_log_status(GSS_C_MECH_CODE, minor);
+
+        ret = EIO;
+        goto done;
+    }
+
+    ret = EOK;
+
+done:
+    gss_release_name(&minor, &name);
+
+    return ret;
+}
+
+static errno_t
+gssapi_handshake(struct gssapi_state *state,
+                 struct cli_protocol *pctx,
+                 const char *keytab,
+                 const char *target,
+                 uint8_t *gss_data,
+                 size_t gss_data_len)
+{
+    OM_uint32 flags = GSS_C_MUTUAL_FLAG;
```

Comment:
Mutual authentication seems reasonable to me. Is it not correct/expected? Do you suggest different flags?

Reply:
I didn't say it was wrong, I said it could use a comment :)

Requiring mutual authentication means that some mechanisms cannot be used here.  If it's important to what you're doing, that's fine - but say why.  In particular, NTLM can't be used (see [gss-ntlmssp](https://github.com/gssapi/gss-ntlmssp/blob/main/src/gss_sec_ctx.c#L119-L129)) - I don't happen to think NTLM *should* be used, but that may or may not be your intent.

Reply:
I think we only care about krb5 now therefore it is fine if NTLM is not supported. Also we control both ends of the communication so we can simply change it if we ever need to. Mutual authentication is not strictly required - we don't exchange any messages we just want to authenticate the user. But it feels right to me to ask for it.